### ISSUE TITLE
ACLs on DRM device not set early enough (logind's DeviceAllow=), breaking GNOME login

### ISSUE THREAD

**Issue by iainlane** [systemd/systemd#14322]
**systemd version the issue has been seen with**
243, 244, master @ da1b880a3a772aa46e023353bdfa4f558286fa5e

**Used distribution**
Ubuntu focal

**Expected behaviour you didn't see**
I expected to be able to log into my GNOME session in a VM (qemu, QXL)

**Unexpected behaviour you saw**
GDM fails to start in Wayland mode. In the log there is

```
Dec 12 10:51:38 focal-vm gnome-shell[911]: Failed to open gpu '/dev/dri/card0': GDBus.Error:org.freedesktop.DBus.Error.AccessDenied: Operation not permitted
Dec 12 10:51:38 focal-vm gnome-shell[911]: Failed to create backend: No GPUs found
Dec 12 10:51:38 focal-vm systemd[861]: gnome-shell-wayland.service: Failed with result 'protocol'.
Dec 12 10:51:38 focal-vm systemd[861]: Failed to start GNOME Shell on Wayland.
Dec 12 10:51:38 focal-vm systemd[861]: Dependency failed for GNOME Shell on Wayland.
Dec 12 10:51:38 focal-vm systemd[861]: Dependency failed for GNOME Wayland Session.
Dec 12 10:51:38 focal-vm systemd[861]: Dependency failed for GNOME Wayland Session (session: gnome-login).
```

This is *not* a race with the device node being created - in Ubuntu we have an `ExecStartPre` script which makes GDM wait for a `master-of-seat` device, so it is there and ready. I verified this by hacking in a script to print out `CanGraphical`, and that is `true`. Then I modified the script to print out `getfacl /dev/dri/card0`, and that shows:

```
Dec 12 10:51:37 focal-vm gdm-wait-for-drm[821]: getfacl: Removing leading '/' from absolute path names
Dec 12 10:51:37 focal-vm gdm-wait-for-drm[821]: # file: dev/dri/card0
Dec 12 10:51:37 focal-vm gdm-wait-for-drm[821]: # owner: root
Dec 12 10:51:37 focal-vm gdm-wait-for-drm[821]: # group: video
Dec 12 10:51:37 focal-vm gdm-wait-for-drm[821]: user::rw-
Dec 12 10:51:37 focal-vm gdm-wait-for-drm[821]: group::rw-
Dec 12 10:51:37 focal-vm gdm-wait-for-drm[821]: other::---
```

The GDM user's ACL isn't there, so uaccess didn't happen yet?

GDM restarts itself in Xorg mode and the ACL is there this time - so this is a race? - but then when I try to log into the user (also Xorg) session, it fails for the same reason and the Xorg log dumped contains:

```
Dec 12 11:22:47 focal-vm /usr/lib/gdm3/gdm-x-session[1417]: (EE) qxl(0): [drm] failed to set drm interface version.
Dec 12 11:22:47 focal-vm /usr/lib/gdm3/gdm-x-session[1417]: (EE) qxl(0): Kernel modesetting setup failed
Dec 12 11:22:47 focal-vm /usr/lib/gdm3/gdm-x-session[1417]: (II) UnloadModule: "qxl"
Dec 12 11:22:47 focal-vm /usr/lib/gdm3/gdm-x-session[1417]: (EE) Screen(s) found, but none have a usable configuration.
Dec 12 11:22:47 focal-vm /usr/lib/gdm3/gdm-x-session[1417]: (EE)
Dec 12 11:22:47 focal-vm /usr/lib/gdm3/gdm-x-session[1417]: Fatal server error:
Dec 12 11:22:47 focal-vm /usr/lib/gdm3/gdm-x-session[1417]: (EE) no screens found(EE)
```

I've bisected this, and the commit which introduced the problem is 9af2820694e1b2d409ed35cf0bca00acab0bdec5 - adding `DeviceAllow` to logind. There is a followup commit 11aa16bb359ded91d68ff5f48fecfafa2a15d3b0. It looks to me like that one is not working here. [Here's a journal log with LogLevel=debug.](https://github.com/systemd/systemd/files/3955551/journal.txt) I'd removed `-q` from `modprobe` so we can see what it's doing. A couple of interesting things I can see in there:

```
Dec 12 11:39:41 focal-vm systemd[1]: Device whitelist pattern "drm" did not match anything.
```

and (but in any event this is after the earlier failure)

```
Dec 12 11:39:42 focal-vm systemd[484]: systemd-logind.service: Executing: /sbin/modprobe -ab drm
Dec 12 11:39:42 focal-vm modprobe[484]: modprobe: ERROR: ../libkmod/libkmod.c:586 kmod_search_moddep() could not open moddep file '/lib/modules/5.3.0-24-generic/modules.dep.bin'
Dec 12 11:39:42 focal-vm modprobe[484]: modprobe: WARNING: Module drm not found in directory /lib/modules/5.3.0-24-generic
```

**Steps to reproduce the problem**

  * Start a QEMU VM (here this is QXL graphics + spice) - I'm using [focal](http://cdimage.ubuntu.com/ubuntu/daily-live/pending/) which currently has this problem out of the box.
  * Try to log in to a GNOME session.
  * You get kicked out back to GDM instead of logged in.

**Comment by poettering** [comment]
So, the DeviceAllow= lines in systemd-logind.service can only be resolved properly if the drm subsystem is already available in the kernel. That's why we do the modprobe -qab in ExecStartPre= which however appears to fail. if it fails, we cannot resolve the subsystem name and thus the DeviceAllow= lines to access drm devices from logind will resolve to nothing.

The question is: why does modprobe fail for you? Why is /lib/modules/5.3.0-24-generic/modules.dep.bin missing?

**Comment by mbiebl** [comment]
apparmor issue?

**Comment by iainlane** [comment]
> Why is /lib/modules/5.3.0-24-generic/modules.dep.bin missing?

It actually is not

```
laney@focal-vm:~$ ls /lib/modules/5.3.0-24-generic/modules.dep.bin
/lib/modules/5.3.0-24-generic/modules.dep.bin
```

dunno what on earth that means.

And after GDM/Xorg comes up:

```
laney@focal-vm:~$ lsmod | grep drm
drm_kms_helper        184320  3 qxl
drm                   491520  5 drm_kms_helper,qxl,ttm
fb_sys_fops            16384  1 drm_kms_helper
syscopyarea            16384  1 drm_kms_helper
sysfillrect            16384  1 drm_kms_helper
sysimgblt              16384  1 drm_kms_helper
```

> apparmor issue?

Interesting theory. I tried with `apparmor=0` on the command line and the issue still persists though, sadly.

**Comment by iainlane** [comment]
I commented out everything to do with lockdown except `DeviceAllow=` in `systemd-login.service`, and that didn't help either.

**Comment by poettering** [comment]
Hmm, but still, the modprobe is failing, and I am pretty sure that's causing the DeviceAllow= line to not be resolvable. You need to figure out why that fails... Maybe it's an early boot race? i.e. modules.dep.bin gets generated at the same time as logind initializes?

(or in other words: I doubt this has anything to do with the device node ACLs as managed by logind and enforced by udev. X/Wayland ask logind to open the devices after all, and logind generally has the privs to open anything it wants, if it wasn't for DeviceAllow=... Hence, track down what happens there... i.e. you could also see if adding drm to modules-load.d/ "fixes" your issue. if it does, you know it's really about the time you are loading that module...

**Comment by iainlane** [comment]
> i.e. you could also see if adding drm to modules-load.d/ "fixes" your issue

Yes, that works. But also, adding this service works - the module gets loaded and I end up in a Wayland greeter, can log in to the session, etc:

```
# /etc/systemd/system/pre-modprobe.service
[Unit]
Description=early modprobe
Before=systemd-logind.service

[Service]
Type=oneshot
ExecStart=/sbin/modprobe -ab --verbose drm

[Install]
WantedBy=systemd-logind.service
```

and modprobe called from systemd-logind.service, happening *after* this, is still complaining about the directory being missing. Ok, let's get systemd-logind.service to tell us what `/lib/` looks like (`ExecStartPre=/bin/ls -l /lib/ /lib/modules/ /lib/modules/5.3.0-24-generic/`):

```
Dec 12 16:16:13 focal-vm ls[492]: /bin/ls: cannot access '/lib/modules/5.3.0-24-generic/': No such file or directory
Dec 12 16:16:13 focal-vm systemd[492]: systemd-logind.service: Executing: /bin/ls -l /lib/ /lib/modules/ /lib/modules/5.3.0-24-generic/
Dec 12 16:16:13 focal-vm ls[492]: /lib/:
Dec 12 16:16:13 focal-vm ls[492]: total 1792
[...]
Dec 12 16:16:13 focal-vm ls[492]: d---------  2 root root      40 Dec 12 16:16 modules
Dec 12 16:16:13 focal-vm ls[492]: /lib/modules/:
Dec 12 16:16:13 focal-vm ls[492]: total 0
```

The permissions on `/lib/modules` are weird. If I add the same `ExecStartPre` to the new `pre-modprobe.service`, it is:

```
Dec 12 16:16:13 focal-vm ls[442]: drwxr-xr-x  4 root root    4096 Dec 10 07:35 modules
```

the timestamp is different also... This is clearly pointing to a hardening feature. Indeed, logind has `ProtectKernelModules=yes`. I've removed that now (not sure if you'd ultimately want to do that), and modprobe fails with:

```
Dec 12 16:46:02 focal-vm modprobe[454]: modprobe: ERROR: could not insert 'drm': Operation not permitted
Dec 12 16:46:02 focal-vm modprobe[454]: insmod /lib/modules/5.3.0-24-generic/kernel/drivers/gpu/drm/drm.ko
```

Ok, that's missing `CAP_SYS_MODULE`. Let's add that.

```
Dec 12 16:46:51 focal-vm systemd[1]: Child 451 (modprobe) died (code=exited, status=0/SUCCESS)
```

Still I end up in Xorg, still with the permission error, and this in the journal:

```
Dec 12 16:46:50 focal-vm systemd[1]: Device whitelist pattern "drm" did not match anything.
...
Dec 12 16:46:50 focal-vm systemd[1]: systemd-logind.service: About to execute: /sbin/modprobe -ab --verbose drm
```

we attempted to resolve for `DeviceAllow` before executing modprobe (that would explain why doing the modprobe in a separate `Before` unit worked).

So a few issues here, any hints as to the right way out?

**Comment by poettering** [comment]
oh, hah! I think this is caused by ProtectKernelModules=yes! that option hides the kmod dir too! Fun!

It probably is sufficient to change this line:

```
ExecStartPre=-/sbin/modprobe -abq drm
```

to:

```
ExecStartPre=-+/sbin/modprobe -abq drm
```

to run that line without any sandboxing.

Does that work?

if so, can you prep a PR?

**Comment by iainlane** [comment]
Oho, I didn't know about `+`! That solves the first two of my problems (dropping the setting and having to add `CAP_SYS_MODULE`). The last issue (ordering problem) remains.

**Comment by poettering** [comment]
hmm, so you are saying that even though the "drm" kmod is successfully loaded the DeviceAllow= line did not resolve to anything?

hmm, if you boot up with init=/bin/sh on the kernel cmdline and execute `modprobe drm ; cat /proc/devices` do you see `drm` listed under character devices?

either the device node major is not actually registered by by drm.ko (i.e. we got it wrong, and a different kmod does that), or it does so asynchronously (i.e. it exits from the kmod initialization function before the major is registered). (or something else that escapes me)

**Comment by iainlane** [comment]
Nah, the module loads properly and causes the major to show up in `/proc/devices`.

It looks to me like the DeviceAllow= is being resolved right *before* we execute the `ExecStartPre=/sbin/modprobe...`, and never tried again. So if `drm.ko` is not loaded at that point, the modprobe does indeed get it loaded, but it's already too late because we concluded it wasn't there, skipped the DeviceAllow=, and never revisited that again for anything else we executed.
